### ISSUE TITLE
NoSuchMethodError: org.assertj.core.api.Assertions.assertThat(Ljava/util/List;)

### ISSUE THREAD

**Issue by rasheedamir** [fabric8io/fabric8#6924]
@jstrachan 

I am trying to debug this issue:

```
java.lang.NoSuchMethodError: org.assertj.core.api.Assertions.assertThat(Ljava/util/List;)Lorg/assertj/core/api/AbstractListAssert;
	at io.fabric8.kubernetes.assertions.KubernetesAssert.deployments(KubernetesAssert.java:97)
```

Here are versions:

```
org.assertj:assertj-core:jar:3.8.0:test
io.fabric8:kubernetes-assertions:jar:2.3.2:test
```

Any thoughts what could be the root cause?

**Comment by jstrachan** [comment]
are you using the fabric8 BOM which defines all the versions of stuff nicely? Or you could try remove the assertj-core dependency as thats imported via kuberentes-assertions anyway?

**Comment by jstrachan** [comment]
fabric8 ships with this version: 2.4.1 https://github.com/fabric8io/fabric8/blob/master/pom.xml#L109

**Comment by jstrachan** [comment]
guess we should try upgrade sometime ;)

**Comment by rasheedamir** [comment]
Ah! I have other tests which needs latest version of assertj and if I use old version then they start failing. Can I help to upgrade it right now?

**Comment by jstrachan** [comment]
that'd be awesome! Wanna fork fabric8io/fabric8 and try upgrade to the latest assertj and see what breaks? There may be some tweaks we need in the kubernetes-assertions to upgrade

**Comment by jstrachan** [comment]
gimme a shout if you hit anything tricky - we do use the code generation with kubernetes-assertions to get the lovely deep navigation assertions so may need to tweak the templates etc

**Comment by rasheedamir** [comment]
OK! we will for it and take a look right away; will ping you on irc if needed any help

**Comment by IAmWaseem** [comment]
@jstrachan I looked briefly at the kubernetes-assertion package as well as the latest assertJ version, and here are my findings:
- You're using 2.4.0 of assertJ which basically returns EA (Type of the object you need to assert) from the toAssert methods
- In 2.8 (latest version) however they changed all of that. They now return the type ObjectAssert<T> instead of the asserted object.
- A brief look at NavigationListAssert showed that extended from ListAssert is now going to be painful since they're added a new layer in between ListAssert and AbstractListAssert and that is FactoryBasedNavigableListAssert, which shows that they now have their own AssertFactory and you can utilize that instead of your own one.
- I did see a lot of places where passing the SELF type as a template is not needed anymore, i'll have to see how they have added that change

Can you tell me if its okay to extract that package and work on that separately if its easier that way? And if there are any changes required for me to run that separately, can you let me know of them as well? Otherwise I can just clone the whole thing and try to upgrade.

Will let you know if I run into some issues in templates

**Comment by jstrachan** [comment]
Damn; sounds like a fair bit has changed so its gonna take a bit of work to refactor kubernetes-assertions. Hopefully the refactoring isn't too hard; but it does sound like a fair amount has changed. 

Am not sure if moving that package to another repo is gonna help much TBH; but you should be able to just build/test that one maven module locally and ignore the other maven modules if you wanna hack on it in a fork?

**Comment by IAmWaseem** [comment]
Yeah it would be great if I could just test that specific module. I'm not exactly sure how to do that since I haven't looked into the project in great detail. If you could give me pointers, I could set that up quickly and see if I can refactor with the addition to the latest assertJ version.

**Comment by IAmWaseem** [comment]
Also I have a question, would upgrading the kubernetes-assertion module to Java 8 cause any issues with dependencies of this module?

**Comment by jstrachan** [comment]
@IAmWaseem lets go for java 8 really; sounds good to me.

if you fork fabric8io/fabric8 then you should be able to type:
```
cd fabric8/components/kubernetes-assertions
mvn install
```
and just keep running/testing that project only. It just happens that the parent `pom.xml` has the versions as maven properties; but other than that you should be able to work exclusively in the `kubernetes-assertions` module

**Comment by IAmWaseem** [comment]
Alright thanks for the tip. Gonna start working on that

P.S I forgot to mention that version 2.8 was the latest version supported for Java 7 and if we move to Java 8 we could upgrade to a better release of assertJ that is 3.8

**Comment by jstrachan** [comment]
going with the latest sounds good to me!
